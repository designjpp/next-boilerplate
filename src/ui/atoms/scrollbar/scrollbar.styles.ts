import { css } from '@emotion/css';
import xw from 'xwind';

const wrapperClassNames = xw`fixed top-0 right-0 z-0 flex h-screen`;
export const wrapperStyles = css`
	${wrapperClassNames}
`;

const scrollbarClassNames = {
	active: xw`rounded-full`,
	base: xw`relative w-3 md:w-4 my-auto h-70vh`,
	theme: xw`rounded-full rounded-tr-none rounded-br-none bg-white-opacity-10`,
};
export const scrollbarStyles = css`
	${scrollbarClassNames.base};
	${scrollbarClassNames.theme};

	&.active {
		${scrollbarClassNames.active};
	}
`;

const thumbClassNames = xw`absolute top-0 right-0 z-0 w-12`;
export const thumbStyles = css`
	${thumbClassNames};
`;

const thumbInnerClassNames = {
	active: xw`rounded-full`,
	base: xw`absolute top-0 right-0 z-0 h-full w-3 md:w-4`,
	theme: xw`bg-blue border-mouse rounded-full rounded-tr-none rounded-br-none`,
	animation: xw`transition-all duration-300 ease-out motion-reduce:transform-none`,
};
export const innerThumbStyles = css`
	${thumbInnerClassNames.base};
	${thumbInnerClassNames.theme};
	${thumbInnerClassNames.animation};

	&.active {
		${thumbInnerClassNames.active};
	}
`;
