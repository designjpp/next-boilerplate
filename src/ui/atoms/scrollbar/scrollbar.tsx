import { cx } from '@emotion/css';
import { useEffect, useMemo, useRef, useState } from 'react';

import { useMouseEffect, useScrollPosition } from 'lib/hooks';

import {
	wrapperStyles,
	scrollbarStyles,
	thumbStyles,
	innerThumbStyles,
} from './scrollbar.styles';

import type { CSSProperties } from 'react';

export const Scrollbar = () => {
	const scrollThumbRef = useRef<HTMLSpanElement>(null);

	const [top, height] = useScrollPosition();
	const noScrollbar = !height;
	const thumbInlineStyles = useMemo<CSSProperties>(
		() => ({
			height: `${height}%`,
			top: `${top}%`,
		}),
		[height, top]
	);

	const [isScrolling, setIsScrolling] = useState(false);
	const onScrollFocus = () => setIsScrolling(true);
	const onScrollBlur = () => setIsScrolling(false);
	useEffect(() => {
		const afterScroll = () => {
			setIsScrolling((prev) => {
				if (prev) {
					scrollThumbRef.current?.blur();
				}

				return prev;
			});
		};

		window.addEventListener('mouseup', afterScroll);

		return () => {
			window.removeEventListener('mouseup', afterScroll);
		};
	}, []);

	const [isExpanded, setIsExpanded] = useState(false);
	useMouseEffect(
		(event) => {
			if (isScrolling) {
				const disableTextSelection = () => {
					window.getSelection()?.removeAllRanges();
				};
				const scroll = () => {
					window.scrollTo({
						top: window.scrollY + event.movementY * 10,
					});
				};

				scroll();
				disableTextSelection();
			} else {
				const handleExpandState = () => {
					const isMouseNear =
						(window.innerWidth - event.x) / window.innerWidth < 0.05;

					if (isMouseNear) {
						setIsExpanded(true);
					} else {
						const movementX = event.movementX;
						setTimeout(() => {
							if (movementX < 0) {
								setIsExpanded(false);
							}
						}, 500);
					}
				};

				handleExpandState();
			}
		},
		[isScrolling]
	);

	return noScrollbar ? null : (
		<span className={wrapperStyles}>
			<span className={cx(scrollbarStyles, isExpanded && 'active')}>
				<span
					ref={scrollThumbRef}
					tabIndex={-1}
					role='scrollbar'
					aria-controls=''
					aria-valuenow={0}
					onTouchStart={onScrollFocus}
					onTouchEnd={onScrollBlur}
					onFocus={onScrollFocus}
					onBlur={onScrollBlur}
					className={thumbStyles}
					style={thumbInlineStyles}
				>
					<span className={cx(innerThumbStyles, isExpanded && 'active')} />
				</span>
			</span>
		</span>
	);
};
