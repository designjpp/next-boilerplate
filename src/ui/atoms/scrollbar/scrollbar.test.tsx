import { render } from '@testing-library/react';
import { axe } from 'jest-axe';
import { create } from 'react-test-renderer';

import { Scrollbar } from './scrollbar';

describe('<Scrollbar>', () => {
	it('should match snapshot', async () => {
		const component = create(<Scrollbar />).toJSON();
		expect(component).toMatchSnapshot();
	});

	it('should be accessible', async () => {
		const { container } = render(<Scrollbar />);
		const results = await axe(container);
		expect(results).toHaveNoViolations();
	});
});
