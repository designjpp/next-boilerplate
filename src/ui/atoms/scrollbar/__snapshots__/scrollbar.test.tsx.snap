// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Scrollbar> should match snapshot 1`] = `null`;
